geopandas/tools/sjoin.py

```diff
@@ -186,6 +180,351 @@ def find_contains(a1, a2):
+            improvement.
+          If possible, you may want to run at least once with
+            max_search_neighbors=None to validate results for your data.
+    search_radius : int or float, default None
+        Restricts search to a certain radius.
+        This can significantly speed up execution.
+        If using a tight search radius, it is recommended that you do not use the
+          max_search_neighbors option as it will not improve performance.
+        If you are not using search_radius, consider using max_search_neighbors
+          to speed up execution.
+    nearest_distances: bool, default False
+        If True, report the distance for each match in new column
+          named "nearest_distances".
+    """
+    # ------------------------------ CHECK INPUTS ------------------------------
+    allowed_hows = ["left", "right", "inner"]
+    if how not in allowed_hows:
+        raise ValueError(
+            '`how` was "%s" but is expected to be in %s' % (how, allowed_hows)
+        )
+
+    _basic_checks(left_df, right_df)
+
+    # ----------------- RENAME INDEXES FOR RTREE COMPATIBILITY -----------------
+    (
+        left_df,
+        right_df,
+        left_index_name,
+        right_index_name,
+        index_left,
+        index_right,
+    ) = _rename_indexes(left_df, right_df, lsuffix, rsuffix)
+
+    # ------------------------ COMPUTE SPATIAL JOIN ----------------------------
+
+    # get spatial index
+    tree_idx = right_df.sindex
+
+    # validate max_search_neighbors and search_radius params
+    if max_search_neighbors is not None:
+        assert max_search_neighbors >= 1 and isinstance(
+            max_search_neighbors, int
+        ), "max_search_neighbors must be an integer and >= 1"
+        # warn about using rtree < 0.9.4 and max_search_neighbors option
+        # see https://github.com/Toblerity/rtree/pull/141
+
+        if [int(i) for i in RTREE_VERSION.split(".")] < [0, 9, 4]:
+            warn(
+                "Using an rtree version < 0.9.4 may cause inconsistent "
+                "results when using max_search_neighbors. Consider using a "
+                "large number or max_search_neighbors=None."
+            )
+    if search_radius is not None:
+        assert search_radius >= 0, "search_radius must be >=0"
+
+    def query_index(geo_in_l):
+        """
+        Queries the spatial index to filter results.
+        """
+        # restric by radius
+        if search_radius is not None:
+            if geo_in_l.geom_type == "Point" and search_radius == 0:
+                # applying .buffer(0) to a point returns an empty geometry
+                geo_search = geo_in_l.coords[0]
+            else:  # polygon or buffer > 0
+                geo_search = geo_in_l.buffer(search_radius).bounds
+            in_radius = list(tree_idx.intersection(geo_search))
+        else:
+            in_radius = right_df.index
+        # find neighbors
+        if max_search_neighbors is not None:
```

Comment:
I would maybe start with `max_search_neighbors` and then used `search_radius` only on the subset identified in the first search. Then the combination would actually bring performance improvement.

Reply:
As explained in the previous comment, I believe this would require rebuilding the spatial index every time, extremely costly.